chain-tx-enclave/enclave-ra/ra-enclave/src/context.rs

```diff
@@ -0,0 +1,228 @@
+    validity_duration: Duration,
+}
+
+impl EnclaveRaContext {
+    /// Creates a new enclave remote attestation context
+    pub fn new(config: &EnclaveRaConfig) -> Result<Self, EnclaveRaContextError> {
+        let sp_ra_client = SpRaClient::connect(&config.sp_addr)?;
+        let validity_duration = Duration::seconds(config.certificate_validity_secs.into());
+
+        Ok(Self {
+            certificate: Default::default(),
+            sp_ra_client,
+            validity_duration,
+        })
+    }
+
+    /// Returns current certificate. If current certificate is no longer valid, then it creates a new one
+    pub fn get_certificate(&self) -> Result<Certificate, EnclaveRaContextError> {
+        let mut certificate = self.certificate.lock().unwrap();
+
+        let needs_creating = match *certificate {
+            None => true,
+            Some(ref certificate) => !certificate.is_valid(self.validity_duration),
+        };
+
+        if needs_creating {
+            let new_certificate = self.create_certificate();
+
+            match new_certificate {
+                Ok(new_certificate) => {
+                    log::info!("Successfully created new certificate for remote attestation");
+                    *certificate = Some(new_certificate);
+                }
+                Err(e) => {
+                    // If the certificate generation fails, we do not crash and keep using the old certificate
+                    // (if available) and client can decide if they want to use the old certificate. Every certificate
+                    // has a 90 days valid duration. If certificate creation fails for 90 days, the enclave itself will
+                    // not serve any client.
+                    log::error!("Failed to create new certificate: {}", e);
+                }
+            }
+        }
+
+        match *certificate {
+            Some(ref certificate) => Ok(certificate.clone()),
+            None => Err(EnclaveRaContextError::CertificateCreationError),
+        }
+    }
+
+    /// Generates attestation report for remote attestation
+    fn get_attestation_report(
+        &self,
+        public_key_hash: [u8; 32],
+    ) -> Result<AttestationReport, EnclaveRaContextError> {
+        // Get target info from SP server
+        let target_info_bytes = self.sp_ra_client.get_target_info()?;
+        let target_info = Targetinfo::try_copy_from(&target_info_bytes)
+            .ok_or_else(|| EnclaveRaContextError::InvalidTargetInfo)?;
+
+        let self_target_info = Targetinfo::from(Report::for_self());
+
+        log::info!("Remote target_info: {:?}", target_info);
+        log::info!("Self target_info: {:?}", self_target_info);
+
+        // Generate enclave report
+        let report = self.get_report(&target_info, public_key_hash)?;
+        let report_bytes: &[u8] = report.as_ref();
+
+        // Get quote and QE report from SP server
+        let quote_result = self.sp_ra_client.get_quote(report_bytes.to_vec())?;
```

Comment:
one should also include + verify nonce: https://github.com/crypto-com/incubator-teaclave/blob/chain/attestation/src/platform.rs#L131

Reply:
Should we create a random nonce each time?

Reply:
yes

Reply:
I'll do it.

Reply:
Done.